python_modules/dagster/dagster/_core/definitions/asset_selection.py

```diff
@@ -1,8 +1,9 @@
 import collections.abc
 import operator
 from abc import ABC, abstractmethod
+from dataclasses import dataclass, replace
```

Comment:
could consider using `pydantic` `dataclass` if we want these to be run time type checked

https://docs.pydantic.dev/latest/concepts/dataclasses/

but since it appears the original `NamedTuple` variants did not have a custom `__new__`  with checks that would be new additional behavior

Reply:
I thought it was a good idea and wanted to do it, but I seems `pydantic` doesn't support `replace`, so I think for this use case we can keep `dataclasses` which does.

Reply:
Would `model_copy` work? https://github.com/pydantic/pydantic/discussions/3352

Reply:
We support both pydantic 1 & 2 and those comments make it seem like `model_copy` is 2 only, something to double check

Reply:
Oh, interesting. That would most likely work. I will implement it before merging.

Reply:
looks like `copy` -> `model_copy` was a name change in 1 -> 2 
https://docs.pydantic.dev/latest/migration/#changes-to-pydanticbasemodel
so I think we would need to handle that

Reply:
we could write a wrapper

Reply:
I implemented `pydantic.BaseModel` and a `wrapper` - the usage of `copy`/`model_copy` requires `pydantic.BaseModel`.

One major con of this implementation is that we are losing the ability make the subclasses `frozen=True` when inheriting from `AssetSelection`. See screenshots at the end of this comments.

Another solution I tried was using the `allow_mutation` config, which raised the same exceptions. 

In the current implementation, the subclasses are mutable, which is something we should avoid if we want to replicate the `NamedTuple` behavior.

I think we have 3 options:

1. Keep the current implementation as is, with the subclasses being mutable.
2. Find a solution to the exception raised when the super class `AssetSelection` is also frozen (see last screenshot of this comment. My initial findings about this are not optimal (see [here](https://stackoverflow.com/questions/63448679/typeerror-init-subclass-takes-no-keyword-arguments)), so if you have any ideas about this one, let me know.
3. Revert to using `dataclasses`, with `replace` and `frozen=true`.

@alangenfeld @sryza WDYT?

When subclasses are frozen but not super class `AssetSelection`:
<img width="908" alt="Screenshot 2024-01-17 at 10 32 57 AM" src="https://github.com/dagster-io/dagster/assets/46797220/44e0eedf-1ffa-46d7-a330-72d76c71f3c9">

When super class `AssetSelection` is also frozen:
<img width="1083" alt="Screenshot 2024-01-17 at 10 33 32 AM" src="https://github.com/dagster-io/dagster/assets/46797220/47b3dd5f-5de9-4492-b7bb-a90c51b4059b">

Reply:
After a few attempts to properly handle inheritance and frozen classes, the code and PR description are updated. @alangenfeld @sryza, lmk what you think!